pkg/driver/node/mounter/pod_unmounter.go

```diff
@@ -230,30 +109,155 @@ func (u *PodUnmounter) CleanupDanglingMounts() error {
 		mpPod, err := u.podWatcher.Get(mpPodName)
 		if err != nil {
 			if apierrors.IsNotFound(err) {
-				klog.V(4).Infof("Mountpoint Pod %s not found, will unmount and delete source folder %s", mpPodName, source)
-
-				// Check mountpoint status and handle special cases
-				isMountPoint, err := u.checkMountpointAndCleanup(source)
-				if !isMountPoint || err != nil {
-					continue
-				}
-
-				// Can only do unmount and remove directory as MP Pod does not exist.
-				if err := u.unmountAndRemoveDir(source); err != nil {
-					klog.Errorf("Failed to cleanup dangling mount %s: %v", source, err)
+				if err := u.unmountAndRemoveMountpointSource(source); err != nil {
+					klog.Errorf("Failed to unmount and remove Mountpoint %q: %v", source, err)
 				}
 				continue
 			}
 
-			klog.Errorf("Failed to check Mountpoint Pod %s existence: %v", mpPodName, err)
-			return err
+			return fmt.Errorf("failed to check existence of Mountpoint Pod %q: %w", mpPodName, err)
 		}
 
-		// Unmount only if Mountpoint Pod is marked for unmounting
-		if value, exists := mpPod.Annotations[mppod.AnnotationNeedsUnmount]; exists && value == "true" {
-			u.unmountSourceForPod(mpPod)
+		u.unmountMountpointPodIfNeeded(mpPod)
+	}
+
+	return nil
+}
+
+// unmountMountpointPodIfNeeded unmounts `mpPod` if and only if annotated with "needs-unmount".
+func (u *PodUnmounter) unmountMountpointPodIfNeeded(mpPod *corev1.Pod) {
+	if value, _ := mpPod.Annotations[mppod.AnnotationNeedsUnmount]; value != "true" {
+		// Not marked for unmount, skip it
+		return
+	}
+
+	unlockMountpointPod := lockMountpointPod(mpPod.Name)
+	defer unlockMountpointPod()
+
+	u.cleanUnmount(mpPod)
+}
+
+// cleanUnmount performs a clean unmount for `mpPod`.
+func (u *PodUnmounter) cleanUnmount(mpPod *corev1.Pod) {
+	source := u.mountpointPodSourcePath(mpPod.Name)
+	podPath := u.podPath(string(mpPod.UID))
+
+	// First, write `mount.exit` file to signal a clean exit to Mountpoint Pod, so it exists with zero code.
+	if err := u.writeExitFile(podPath); err != nil {
```

Comment:
Could this fail to write exit file if we already unmounted and deleted MPPod in previous invocation? 

I.e. should we check if `source` is mountpoint before writing exit file?

Reply:
It might fail, but I think it wouldn't be a fatal failure. We already know Mountpoint Pod exists here, and we'd need to signal a clean exit to the Mountpoint Pod even it's not mounted. I also updated `aws-s3-csi-mounter` to check this file even if Mountpoint Pod is not mounted.

Reply:
> We already know Mountpoint Pod exists here

It might not exist as subsequent invocations wait for MP Pod lock release from previous invocation. Although, the chance of MP Pod not existing right away is very low as it takes take to cleanup. So, I am ok to keeping it as it is for now.

Reply:
Yeah at the worst case it should just emit an error and not try it again on the next run